### ISSUE TITLE
ImageData from chunk without rendering

### ISSUE THREAD

**Issue by Pryanishnik** [w3c/webcodecs#343]
Is there any way to get raw ImageData in some format from the chunk we get in WebCodecs? We're working on real time video analytics and the neccessary part is to get original video colors, that is impossible (as I know) due to built-in codecs and canvas color correction thing. Is there any way to extract raw image from a chunk?

**Comment by Pryanishnik** [comment]
@padenot I don't have VideoFrame.copyTo in runtime, but I have VideoFrame.readInto that seems to be doing the same. Yes, I get some buffer array, so now to get an actual RGB, YUV, whatever color format, I have to decode this array in some way?

**Comment by padenot** [comment]
Those formats are described here: https://w3c.github.io/webcodecs/#pixel-format, you can find the format for a frame in its `format` attribute. `readInto` is the old name yes.

**Comment by Pryanishnik** [comment]
@padenot Thank you for the help, manage to decode YUV to RGB. But what I see, that colors are different comparing to opening the same frame in OpenCV i.e. Also, Red (255,0,0) is corrected to some other color. Is VideoFrame data already color corrected with codecs/chrome?

**Comment by dalecurtis** [comment]
In Chrome's implementaiton, `copyTo()` may have to invoke a GPU readback to a known pixel format so some conversion may occur. What is the source of your VideoFrame objects?

**Comment by Pryanishnik** [comment]
@dalecurtis I took as an example  the webcamera in a worker sample, and use OBS virtual camera with static jpeg image as a "video".  When I run webcamera sample and put the same stream to opencv, the rgb results are different. for some picture regions the difference is small, for ome other - pretty significant. I get I420 YUV data, and then convert it to RGB. I think that conversion is right, coz I see that for some pixels difference is really small (rounded it's 1)

**Comment by dalecurtis** [comment]
Camera capture may happen through multiple operating modes depending on the device, os, and drivers. E.g., a camera may be opened in MJPEG or NV12 depending on the camera's format support. If this is Windows you might try running Chrome with `--enable-features=MediaFoundationD3D11VideoCapture`

**Comment by Pryanishnik** [comment]
@dalecurtis Didn't know about the flag, will try, thank you. But this flag is not the best solution for third party users, who actually use some web portal:) or can it be forced in some way (i.e. like trial token for webcodecs)? Also, did I understand you right that the same stream (As I described before, I use absolutely same OBS virtual camera with absolutely same picture in a focus, and run two virtual camera "consumers" in parallel, so data going from OBS virtual camera to browser and opencv client should be absolutely same) can be still different in how browser actually gets data from virtual camera comparing to other clients?

**Comment by dalecurtis** [comment]
We're in the process of launching that everywhere, I just suggested it as a means for seeing if it affects your results locally. Even with that flag though some cameras, OS, and drivers may not offer the pure passthrough experience you're looking for.

In specific regards to the virtual camera, if it supports NV12 then you should get a zero copy passthrough. Chrome avoids capturing in RGB since that's inefficient for encoders and display generally, so if you're providing an RGB source it will go through conversion. I don't know how the virtual camera works, but a 4:2:0 JPEG seems like the best source.

**Comment by sandersdan** [comment]
Ultimately this is a difference in color space interpretation or conversion. The raw bytes from `frame.copyTo()` are in the color space `frame.colorSpace`, although at present we can't represent all the options there and it may just be empty. If you render to `<canvas>` though it will be converted to sRGB and you can compare those pixels between applications.

As Dale mentioned, there may be a conversion happening in Chrome. It would be a bug in our WebCodecs implementation if the `colorSpace` was non-empty and incorrect or if the <canvas> rendering was incorrect. Without the `colorSpace` it's not possible to say whether the raw bytes are correct or incorrect.

> Also, Red (255,0,0) is corrected to some other color.

> I see that for some pixels difference is really small (rounded it's 1)

These sound to me like BT 601 vs BT 709 color primaries.

**Comment by Pryanishnik** [comment]
@sandersdan thank you for reply! As I see, VideoFrame comes with format: "I420". After that I get data from video frame(with readInto coz in 92 chrome its still readInto), and convert YUV to RGB for a canvas. I draw frame to canvas and see that my decoding is right (obviously on a photo I don't see 1 point in red difference:)).

Regarding BT601 and BT709, do you mean that i.e. Webcodecs interpret data from camera as BT601 and other not browser client as BT709 (BTs are put as an example, not as a fact:)) ?

**Comment by sandersdan** [comment]
I mean that it's possible that a conversion in the capture path converted the image from its original representation into BT 601 or BT 709, which could differ from the original source. As an example, libyuv RGB to YUV conversions always produce 601, even though that's rarely what you would want.

It's also possible that at some point in the pipeline the color space information was incorrectly overwritten, which would be a bug. There isn't enough information here to know if something like that has happened.

**Comment by Pryanishnik** [comment]
@sandersdan thank you for the help! I really don't compare browser colors with the original image, I do compare with the same frame in browser and opencv VideoCapture thing, so actually Mat Object received from the same virtual camera stream. So I think that might be that OpenCV use different primaries comparing to Webcodecs. Will try to figure it out, thank you! Also will take a look at Canary behaviour

**Comment by Pryanishnik** [comment]
@sandersdan Aslo, summarizing all things, on the stage of VideoFrame returned from StreanReader, data is not color corrected with ICC or some other factors yes, right? So difference in colors is because of some mismatches/different intepretations/primaries etc (most likely)?

**Comment by sandersdan** [comment]
ICC profiles are not something we would support for capture or video rendering, although in this case I doubt they would be supported by the underlying platform anyway. In some configurations Chrome does render its window according to the display's color profile, but this shouldn't affect your work unless you are using screenshots.

**Comment by Pryanishnik** [comment]
@sandersdan Yes, that's exactly why we want to avoid drawing frames from video to canvas and then getting ImageData. Thanks for the clear answer!
